better error for "ansible-galaxy collection verify" if there's no MANIFEST.json

s-hertel opened the pull request:
##### SUMMARY
The MANIFEST.json may have been deleted from the installed collection or the collection may not have been installed via normal means

The error message is misleading right now (e.g. `openstack.cloud has the version '*' but is being compared to '1.0.0'`) since an installed collection could never have an undefined version. Built collections do not have a galaxy.yml for finding the version and only built collections should be install-able for verification against a server.

##### ISSUE TYPE
- Bugfix Pull Request

s-hertel committed:
Add a better error for "ansible-galaxy verify" if the MANIFEST.json has been deleted from the installed collection or if the collection hasn't been installed via normal means

s-hertel commented on lib/ansible/galaxy/collection.py:
```diff
@@ -578,6 +578,10 @@ def verify_collections(collections, search_paths, apis, validate_certs, ignore_e
                             break
                     if local_collection is None:
                         raise AnsibleError(message='Collection %s is not installed in any of the collection paths.' % collection_name)
+                    if local_collection.latest_version == '*':
```
It's technically possible that someone could `ansible-galaxy collection init` and then edit the galaxy.yml version to be "*", then build and publish. I feel like no one should do that, but if I should handle it as a valid, specific version, then the check/error should happen after remote_collection is fetch and be something like this:
```python
if local_collection.latest_version == '*' and remote_collection.latest_version != '*':
    msg = 'Collection %s does not appear to have the MANIFEST.json. ' % collection_name
    msg += 'A MANIFEST.json is expected if the collection has been built and installed via ansible-galaxy.'
    raise AnsibleError(message=msg)
```
Other opinions welcome.

sivel replied:
I'm not sure they could actually publish to Galaxy. It would reject that version as not being semver.

s-hertel replied:
Okay, gtk. I only tried to build with the version "*" and was surprised it worked.

s-hertel committed:
Fix unit tests for the remote collection

If there's something wrong with the local collection's version it will fail before the remote collection is sought

s-hertel committed:
Add a test for the new error msg

samdoran commented on lib/ansible/galaxy/collection.py:
```diff
@@ -578,6 +578,10 @@ def verify_collections(collections, search_paths, apis, validate_certs, ignore_e
+                    if local_collection.latest_version == '*':
+                        msg = 'Collection %s does not appear to have the MANIFEST.json. ' % collection_name
+                        msg += 'A MANIFEST.json is expected if the collection has been built and installed via ansible-galaxy.'
+                        raise AnsibleError(message=msg)
```
My only minor problem is that now we show both a warning and an error that essentially say the same thing.
```
[WARNING]: Collection at '/Users/sdoran/.ansible/collections/ansible_collections/openstack/cloud' does not have a MANIFEST.json file,
cannot detect version.
ERROR! Collection openstack.cloud does not appear to have the MANIFEST.json. A MANIFEST.json is expected if the collection has been built and installed via ansible-galaxy.
```

Since these are coming from two different places, there isn't currently a clear way to suppress the warning because we may later raise an error. That's my only minor issue with this change.

s-hertel replied:
I wondered about that too. I added the exception to halt execution (unless ignore errors is used)... but looking at this again, I could instead check os.path.isfile for the MANIFEST.json before calling from_path() and then we'd get rid of the duplicate.

s-hertel replied:
That should be fixed in 470dfbe776cb51be353154f7697555cebc995707.

samdoran replied:
Awesome. This is much better now.

s-hertel committed:
Prevent the duplicate warning

Mock the new isfile call where needed

mkrizek commented on lib/ansible/galaxy/collection.py:
```diff
@@ -574,6 +574,11 @@ def verify_collections(collections, search_paths, apis, validate_certs, ignore_e
                     for search_path in search_paths:
                         b_search_path = to_bytes(os.path.join(search_path, namespace, name), errors='surrogate_or_strict')
                         if os.path.isdir(b_search_path):
+                            if not os.path.isfile(os.path.join(to_text(b_search_path), 'MANIFEST.json')):
```
We should use `errors=surrogate_or_strict` when dealing with file paths:
```suggestion
                            if not os.path.isfile(os.path.join(to_text(b_search_path, errors='surrogate_or_strict'), 'MANIFEST.json')):
```

s-hertel replied:
Oh, oops. Thanks!

s-hertel committed:
Update lib/ansible/galaxy/collection.py

Co-Authored-By: Martin Krizek <martin.krizek@gmail.com>

geerlingguy commented:
Related? https://github.com/ansible/ansible/issues/69311

s-hertel commented:
Yes. I commented on that issue with this PR. https://github.com/ansible/ansible/pull/68925 is the main fix though.
